governance/program/src/state/proposal.rs

```diff
@@ -473,13 +473,35 @@ impl ProposalV2 {
             get_min_vote_threshold_weight(&config.vote_threshold_percentage, max_vote_weight)
                 .unwrap();
 
-        if yes_vote_weight >= min_vote_threshold_weight
-            && yes_vote_weight > (max_vote_weight - yes_vote_weight)
-        {
-            yes_option.vote_result = OptionVoteResult::Succeeded;
-            return Some(ProposalState::Succeeded);
-        } else if deny_vote_weight > (max_vote_weight - min_vote_threshold_weight)
-            || deny_vote_weight >= (max_vote_weight - deny_vote_weight)
+        match config.vote_tipping {
+            VoteTipping::Disabled => {}
+            VoteTipping::Strict => {
+                if yes_vote_weight >= min_vote_threshold_weight
+                    && yes_vote_weight > (max_vote_weight - yes_vote_weight)
+                {
+                    yes_option.vote_result = OptionVoteResult::Succeeded;
+                    return Some(ProposalState::Succeeded);
+                }
+            }
+            VoteTipping::Early => {
+                solana_program::msg!(
+                    "{} {} {}",
+                    yes_vote_weight,
+                    deny_vote_weight,
+                    min_vote_threshold_weight
+                );
+                if yes_vote_weight >= min_vote_threshold_weight
+                    && yes_vote_weight > deny_vote_weight
+                {
+                    yes_option.vote_result = OptionVoteResult::Succeeded;
+                    return Some(ProposalState::Succeeded);
+                }
+            }
+        }
+
+        if !matches!(config.vote_tipping, VoteTipping::Disabled)
```

Comment:
nit: `config.vote_tipping != VoteTipping::Disabled` ?
Can you add comments what happens here please? That we never tip early on `Deny` votes and it's always `Strict` mode for `Deny`